references/obj_detection/train_pytorch.py

```diff
@@ -0,0 +1,204 @@
+# Copyright (C) 2021, Mindee.
+
+# This program is licensed under the Apache License version 2.
+# See LICENSE or go to <https://www.apache.org/licenses/LICENSE-2.0.txt> for full license details.
+
+import os
+
+os.environ['USE_TORCH'] = '1'
+
+import argparse
+import json
+
+import random
+import time
+
+import cv2
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from doctr.transforms.functional.pytorch import rotate
+from sklearn.model_selection import train_test_split
+from torch.utils.data import Dataset, DataLoader
+from tqdm import tqdm
+
+from model import faster_model
+from utilities import *
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+
+class Train_dataset(Dataset):
+    """
+    Args:
+        root_dir: location of image folder
+        label_path: location of labels/annotations
+    Returns:
+        Iterable dataset accessible with integer indices.
+    """
+
+    def __init__(self, root_dir: str, label_path: str, val=False):
+        self.root_dir = root_dir
+        self.root = d_train if not val else d_val
+        self.label_path = label_path
+        with open(self.label_path) as f:
+            self.my_json = dict(json.load(f))
+
+    def __len__(self):
+        return len(self.root)
+
+    def __getitem__(self, idx):
+        target_ = label_formatter({self.root[idx]: self.my_json[self.root[idx]]}, width, height)
+        boxes = np.array(target_[self.root[idx]]['boxes'])
+        image = torch.tensor(image, dtype=torch.float32).permute(2, 0, 1)
+        bbox = [torch.tensor(p, dtype=torch.float32) for p in boxes]
+        return {"image": image,
+                "target": {"boxes": bbox, "labels": target_[self.root[idx]]['labels']}}
+
+class FocalLoss(nn.Module):
+    """
+    Focal Loss to be replaced with the BCE loss from the Faster Rcnn
+    """
+
+    def __init__(self, alpha=0.25, gamma=2, ):
+        super(FocalLoss, self).__init__()
+        self.alpha = alpha
+        self.gamma = gamma
+
+    def forward(self, inputs):
+        loss = [l for l in inputs.values()]
+        pt = torch.exp(-loss[0])
+        F_loss_1 = self.alpha * (1 - pt) ** self.gamma * loss[0]
+        pt_2 = torch.exp(-loss[2])
+        F_loss_2 = self.alpha * (1 - pt_2) ** self.gamma * loss[2]
+        return sum([F_loss_1, loss[1], F_loss_2, loss[3]])
+
+
+fl = FocalLoss()
```

Comment:
Have you noticed any significant improvement by changing the loss? (otherwise let's keep it simple)

Reply:
I am yet to compare the impact on the metrics , but based on the past loss results, I do remember that it indeed had a good impact on the convergence of losses. I will update you  soon with the performance results in both the cases(with and without focal loss)

Reply:
Thanks, Val metrics much better without Focal Loss. It will be removed .

Reply:
Perfect, a shorter boilerplate then :+1: